File: src/lib/types/ai-config.ts
/**
 * AI 配置类型定义
 * 与后端字段完全一致，确保前后端统一
 */

/**
 * 厂商类型枚举
 */
export enum ProviderVendor {
  OPENAI_COMPATIBLE = "openai_compatible",
  ANTHROPIC_COMPATIBLE = "anthropic_compatible",
  GOOGLE_COMPATIBLE = "google_compatible",
  ZHIPU_COMPATIBLE = "zhipu_compatible",
}

/**
 * 接口风格枚举
 */
export enum ApiStyle {
  OPENAI_RESPONSES = "openai_responses",
  OPENAI_CHAT_COMPLETIONS = "openai_chat_completions",
  ANTHROPIC_MESSAGES = "anthropic_messages",
  GEMINI_GENERATE_CONTENT = "gemini_generate_content",
  ZHIPU_CHAT_COMPLETIONS = "zhipu_chat_completions",
  ZHIPU_OCR = "zhipu_ocr",
}

/**
 * PDF 策略枚举
 */
export enum PDFStrategy {
  DIRECT_PDF = "direct_pdf",
  CONVERT_TO_IMAGES = "convert_to_images",
}

/**
 * AI 配置完整模型
 */
export interface AIConfig {
  id: number;
  name: string;
  provider_vendor: ProviderVendor;
  api_style: ApiStyle;
  base_url: string;
  model_name: string;
  api_key_masked: string; // 脱敏后的 API Key
  has_api_key: boolean; // 是否已设置 API Key
  pdf_strategy: PDFStrategy; // PDF 处理策略
  ocr_chat_model?: string; // OCR 模式下结构化提取用的 chat 模型
  timeout: number;
  temperature: number;
  max_tokens: number;
  max_concurrency: number;
  supports_image_input: boolean;
  supports_pdf_file_input: boolean;
  supports_file_id: boolean;
  supports_file_url: boolean;
  requires_files_api: boolean;
  supports_structured_json: boolean;
  enabled: boolean;
  active: boolean;
  created_at: string;
  updated_at: string;
}

/**
 * 创建 AI 配置请求
 */
export interface AIConfigCreateRequest {
  name: string;
  provider_vendor: ProviderVendor;
  api_style: ApiStyle;
  base_url: string;
  model_name: string;
  api_key: string;
  pdf_strategy?: PDFStrategy;
  ocr_chat_model?: string;
  timeout?: number;
  temperature?: number;
  max_tokens?: number;
  max_concurrency?: number;
  supports_image_input?: boolean;
  supports_pdf_file_input?: boolean;
  supports_file_id?: boolean;
  supports_file_url?: boolean;
  requires_files_api?: boolean;
  supports_structured_json?: boolean;
  enabled?: boolean;
}

/**
 * 更新 AI 配置请求
 */
export interface AIConfigUpdateRequest {
  name?: string;
  provider_vendor?: ProviderVendor;
  api_style?: ApiStyle;
  base_url?: string;
  model_name?: string;
  api_key?: string;
  pdf_strategy?: PDFStrategy;
  ocr_chat_model?: string;
  timeout?: number;
  temperature?: number;
  max_tokens?: number;
  max_concurrency?: number;
  supports_image_input?: boolean;
  supports_pdf_file_input?: boolean;
  supports_file_id?: boolean;
  supports_file_url?: boolean;
  requires_files_api?: boolean;
  supports_structured_json?: boolean;
  enabled?: boolean;
}

/**
 * 测试连接响应
 */
export interface DiagnosticStep {
  step: string;
  url: string;
  status_code: number | null;
  latency_ms: number | null;
  success: boolean;
  detail: string;
}

export interface AIConfigTestResponse {
  success: boolean;
  message: string;
  latency_ms?: number;
  request_path?: string;
  status_code?: number;
  diagnostic_steps: DiagnosticStep[];
}

/**
 * 样本测试模式枚举
 */
export enum SampleTestMode {
  IMAGE_SINGLE = "image_single",
  PDF_COMPARE = "pdf_compare",
}

/**
 * 样本测试结果
 */
export interface SampleTestResult {
  success: boolean;
  strategy?: string;
  latency_ms?: number;
  request_path?: string;
  execution_path?: string;
  status_code?: number;
  error_message?: string;
  raw_response?: Record<string, any>;
  structured_result?: Record<string, any>;
  diagnostic_steps?: DiagnosticStep[];
  field_completeness?: {
    required_fields_total: number;
    recognized_fields_count: number;
    missing_fields: string[];
    completeness_rate: number;
  };
  total_pages?: number;
  all_page_results?: Record<string, any>[];
  best_page_index?: number;
}

/**
 * 样本测试响应
 */
export interface SampleTestResponse {
  test_mode: SampleTestMode;
  file_name: string;
  file_type: string;
  invoice_type: string;
  result?: SampleTestResult;
  direct_pdf_result?: SampleTestResult;
  convert_to_images_result?: SampleTestResult;
  recommended_strategy?: string;
  recommendation_reason?: string;
}

/**
 * 更新 PDF 策略请求
 */
export interface UpdatePDFStrategyRequest {
  pdf_strategy: PDFStrategy;
}

/**
 * 厂商类型显示名称映射
 */
export const PROVIDER_VENDOR_LABELS: Record<ProviderVendor, string> = {
  [ProviderVendor.OPENAI_COMPATIBLE]: "OpenAI 兼容",
  [ProviderVendor.ANTHROPIC_COMPATIBLE]: "Anthropic 兼容",
  [ProviderVendor.GOOGLE_COMPATIBLE]: "Google Gemini",
  [ProviderVendor.ZHIPU_COMPATIBLE]: "智谱 GLM",
};

/**
 * 接口风格显示名称映射
 */
export const API_STYLE_LABELS: Record<ApiStyle, string> = {
  [ApiStyle.OPENAI_RESPONSES]: "OpenAI Responses",
  [ApiStyle.OPENAI_CHAT_COMPLETIONS]: "OpenAI Chat Completions",
  [ApiStyle.ANTHROPIC_MESSAGES]: "Anthropic Messages",
  [ApiStyle.GEMINI_GENERATE_CONTENT]: "Gemini generateContent",
  [ApiStyle.ZHIPU_CHAT_COMPLETIONS]: "智谱 Chat Completions",
  [ApiStyle.ZHIPU_OCR]: "智谱 OCR（文件上传+结构化）",
};

/**
 * PDF 策略显示名称映射
 */
export const PDF_STRATEGY_LABELS: Record<PDFStrategy, string> = {
  [PDFStrategy.DIRECT_PDF]: "直接传 PDF",
  [PDFStrategy.CONVERT_TO_IMAGES]: "转图片后识别",
};

/**
 * 根据厂商类型获取支持的接口风格
 */
export function getSupportedApiStyles(vendor: ProviderVendor): ApiStyle[] {
  switch (vendor) {
    case ProviderVendor.OPENAI_COMPATIBLE:
      return [ApiStyle.OPENAI_CHAT_COMPLETIONS, ApiStyle.OPENAI_RESPONSES];
    case ProviderVendor.ANTHROPIC_COMPATIBLE:
      return [ApiStyle.ANTHROPIC_MESSAGES];
    case ProviderVendor.GOOGLE_COMPATIBLE:
      return [ApiStyle.GEMINI_GENERATE_CONTENT];
    case ProviderVendor.ZHIPU_COMPATIBLE:
      return [ApiStyle.ZHIPU_CHAT_COMPLETIONS, ApiStyle.ZHIPU_OCR];
    default:
      return [];
  }
}

/**
 * 脱敏 API Key
 */
export function maskApiKey(apiKey: string): string {
  if (!apiKey || apiKey.length <= 8) {
    return "***";
  }
  return `${apiKey.slice(0, 8)}...`;
}

File: backend/app/schemas/ai_config.py
from pydantic import BaseModel, Field, field_validator
from typing import Optional, List
from datetime import datetime
from enum import Enum


class ProviderVendorEnum(str, Enum):
    """AI 提供商厂商枚举"""
    OPENAI_COMPATIBLE = "openai_compatible"
    ANTHROPIC_COMPATIBLE = "anthropic_compatible"
    GOOGLE_COMPATIBLE = "google_compatible"
    ZHIPU_COMPATIBLE = "zhipu_compatible"


class APIStyleEnum(str, Enum):
    """API 风格枚举"""
    OPENAI_RESPONSES = "openai_responses"
    OPENAI_CHAT_COMPLETIONS = "openai_chat_completions"
    ANTHROPIC_MESSAGES = "anthropic_messages"
    GEMINI_GENERATE_CONTENT = "gemini_generate_content"
    ZHIPU_CHAT_COMPLETIONS = "zhipu_chat_completions"
    ZHIPU_OCR = "zhipu_ocr"


class PDFStrategyEnum(str, Enum):
    """PDF 处理策略枚举"""
    DIRECT_PDF = "direct_pdf"
    CONVERT_TO_IMAGES = "convert_to_images"


# ==================== 请求 Schema ====================

class AIConfigCreate(BaseModel):
    """创建 AI 配置请求"""
    name: str = Field(..., min_length=1, max_length=100, description="配置名称")
    provider_vendor: ProviderVendorEnum = Field(..., description="提供商厂商")
    api_style: APIStyleEnum = Field(..., description="API 风格")
    base_url: str = Field(..., min_length=1, max_length=500, description="API 基础 URL")
    model_name: str = Field(..., min_length=1, max_length=100, description="模型名称")
    api_key: str = Field(..., min_length=1, description="API 密钥")
    timeout: int = Field(default=60, ge=10, le=300, description="超时时间（秒）")
    temperature: float = Field(default=0.1, ge=0.0, le=2.0, description="温度参数")
    max_tokens: int = Field(default=4000, ge=100, le=100000, description="最大 token 数")
    max_concurrency: int = Field(default=1, ge=1, le=100, description="最大并发数")

    # 提供商能力
    supports_image_input: bool = Field(default=True, description="是否支持图片输入")
    supports_pdf_file_input: bool = Field(default=False, description="是否支持 PDF 文件输入")
    supports_file_id: bool = Field(default=False, description="是否支持文件 ID")
    supports_file_url: bool = Field(default=False, description="是否支持文件 URL")
    requires_files_api: bool = Field(default=False, description="是否需要先调用 files API")
    supports_structured_json: bool = Field(default=False, description="是否支持结构化 JSON 输出")

    # PDF 处理策略
    pdf_strategy: PDFStrategyEnum = Field(default=PDFStrategyEnum.CONVERT_TO_IMAGES, description="PDF 处理策略")

    # 智谱 OCR 专用
    ocr_chat_model: Optional[str] = Field(None, max_length=100, description="OCR 模式下结构化提取用的 chat 模型")

    # 状态
    enabled: bool = Field(default=True, description="是否启用")

    @field_validator('name')
    @classmethod
    def validate_name(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("配置名称不能为空")
        return v.strip()

    @field_validator('base_url')
    @classmethod
    def validate_base_url(cls, v: str) -> str:
        v = v.strip()
        if not v.startswith(('http://', 'https://')):
            raise ValueError("API 基础 URL 必须以 http:// 或 https:// 开头")
        return v.rstrip('/')


class AIConfigUpdate(BaseModel):
    """更新 AI 配置请求"""
    name: Optional[str] = Field(None, min_length=1, max_length=100, description="配置名称")
    provider_vendor: Optional[ProviderVendorEnum] = Field(None, description="提供商厂商")
    api_style: Optional[APIStyleEnum] = Field(None, description="API 风格")
    base_url: Optional[str] = Field(None, min_length=1, max_length=500, description="API 基础 URL")
    model_name: Optional[str] = Field(None, min_length=1, max_length=100, description="模型名称")
    api_key: Optional[str] = Field(None, min_length=1, description="API 密钥")
    timeout: Optional[int] = Field(None, ge=10, le=300, description="超时时间（秒）")
    temperature: Optional[float] = Field(None, ge=0.0, le=2.0, description="温度参数")
    max_tokens: Optional[int] = Field(None, ge=100, le=100000, description="最大 token 数")
    max_concurrency: Optional[int] = Field(None, ge=1, le=100, description="最大并发数")

    # 提供商能力
    supports_image_input: Optional[bool] = Field(None, description="是否支持图片输入")
    supports_pdf_file_input: Optional[bool] = Field(None, description="是否支持 PDF 文件输入")
    supports_file_id: Optional[bool] = Field(None, description="是否支持文件 ID")
    supports_file_url: Optional[bool] = Field(None, description="是否支持文件 URL")
    requires_files_api: Optional[bool] = Field(None, description="是否需要先调用 files API")
    supports_structured_json: Optional[bool] = Field(None, description="是否支持结构化 JSON 输出")

    # PDF 处理策略
    pdf_strategy: Optional[PDFStrategyEnum] = Field(None, description="PDF 处理策略")

    # 智谱 OCR 专用
    ocr_chat_model: Optional[str] = Field(None, max_length=100, description="OCR 模式下结构化提取用的 chat 模型")

    # 状态
    enabled: Optional[bool] = Field(None, description="是否启用")

    @field_validator('name')
    @classmethod
    def validate_name(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and not v.strip():
            raise ValueError("配置名称不能为空")
        return v.strip() if v else None

    @field_validator('base_url')
    @classmethod
    def validate_base_url(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            v = v.strip()
            if not v.startswith(('http://', 'https://')):
                raise ValueError("API 基础 URL 必须以 http:// 或 https:// 开头")
            return v.rstrip('/')
        return None


class AIConfigTestRequest(BaseModel):
    """测试 AI 配置请求（已废弃，改用 config_id）"""
    provider_vendor: ProviderVendorEnum = Field(..., description="提供商厂商")
    api_style: APIStyleEnum = Field(..., description="API 风格")
    base_url: str = Field(..., description="API 基础 URL")
    model_name: str = Field(..., description="模型名称")
    api_key: str = Field(..., description="API 密钥")
    timeout: int = Field(default=60, description="超时时间（秒）")


# ==================== 响应 Schema ====================

class AIConfigDetail(BaseModel):
    """AI 配置详情"""
    id: int
    name: str
    provider_vendor: ProviderVendorEnum
    api_style: APIStyleEnum
    base_url: str
    model_name: str
    api_key_masked: str  # 脱敏后的 API Key
    has_api_key: bool  # 是否已设置 API Key
    timeout: int
    temperature: float
    max_tokens: int
    max_concurrency: int
    supports_image_input: bool
    supports_pdf_file_input: bool
    supports_file_id: bool
    supports_file_url: bool
    requires_files_api: bool
    supports_structured_json: bool
    pdf_strategy: PDFStrategyEnum
    ocr_chat_model: Optional[str] = None
    enabled: bool
    active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class AIConfigListItem(BaseModel):
    """AI 配置列表项"""
    id: int
    name: str
    provider_vendor: ProviderVendorEnum
    api_style: APIStyleEnum
    base_url: str
    model_name: str
    api_key_masked: str  # 脱敏后的 API Key
    has_api_key: bool  # 是否已设置 API Key
    timeout: int
    temperature: float
    max_tokens: int
    max_concurrency: int
    supports_image_input: bool
    supports_pdf_file_input: bool
    supports_file_id: bool
    supports_file_url: bool
    requires_files_api: bool
    supports_structured_json: bool
    pdf_strategy: PDFStrategyEnum
    ocr_chat_model: Optional[str] = None
    enabled: bool
    active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class DiagnosticStep(BaseModel):
    """诊断步骤"""
    step: str
    url: str = ""
    status_code: Optional[int] = None
    latency_ms: Optional[int] = None
    success: bool = False
    detail: str = ""


class AIConfigTestResponse(BaseModel):
    """测试 AI 配置响应"""
    success: bool
    message: str
    latency_ms: Optional[int] = None
    request_path: Optional[str] = None
    status_code: Optional[int] = None
    diagnostic_steps: list[DiagnosticStep] = []

